File: ocaml/lib/day1.ml
let modulo n m = ((n mod m) + m) mod m
let num_at n s = String.to_seq s |> Seq.drop n |> String.of_seq |> int_of_string

let part1 input =
  let _, count =
    List.fold_left
      (fun (dial, count) it ->
        let dial =
          match (String.get it 0, num_at 1 it) with
          | 'L', n -> modulo (dial - n) 100
          | 'R', n -> modulo (dial + n) 100
          | _ -> assert false
        in
        if dial = 0 then (dial, count + 1) else (dial, count))
      (50, 0) input
  in
  count

let part2 input =
  let rotate (dial, count) s =
    let increment =
      match String.get s 0 with 'L' -> -1 | 'R' -> 1 | _ -> assert false
    in
    let clicks = num_at 1 s in
    List.init clicks Fun.id
    |> List.fold_left
         (fun (dial, count) _ ->
           let new_dial = modulo (dial + increment) 100 in
           (new_dial, count + if new_dial = 0 then 1 else 0))
         (dial, count)
  in
  let _, count = List.fold_left rotate (50, 0) input in
  count

File: python/day1.py
def turn_dial(dir: str, prev: int, clicks: int) -> int:
    if dir == 'R':
        return (prev + clicks) % 100
    elif dir == 'L':
        return (prev - clicks) % 100
    else:
        raise ValueError("Invalid dir!")


def part1(lines: list[str]) -> int:
    dial = 50
    count = 0
    for line in lines:
        dir = line[0]
        num = int(line[1:])
        dial = turn_dial(dir, dial, num)
        if dial == 0:
            count += 1

    return count

def part2(lines) -> int:
    dial = 50
    count = 0
    for line in lines:
        dir = line[0]
        num = int(line[1:])
        for _ in range(num):
            dial = turn_dial(dir, dial, 1)
            if dial == 0:
                count += 1
    return count
